### PR TITLE
feat: Screenshot

### LINKED ISSUE #707: Request: screenshot
### Priority

Missing Content

### Area

- [ ] Data
- [X] Source
- [ ] Docker
- [ ] Other

### What is missing?

Would it be possible to add the screenshot function?
It would be interesting to make it compatible with this.

TFS
https://github.com/otland/forgottenserver/pull/4459

canary
https://github.com/opentibiabr/canary/pull/2116/files

Or a simple module that automatically takes a screenshot when you level up or when you die, for example.
https://github.com/EgzoT/-OTClient-Mod-game_autoscreenshot

image

Preferably in the same folder as the client. Not in AppData. (For users with basic computer skills.)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kokekanon**
# Description

there was a packet error in 13.10  was Screenshot packets  0x75 (117), i tried to solve the issues of #707

image

I have doubts whether it is g_game or LocalPlayer.

the image is delayed between 6 to 2 seconds
```
Attempting screenshot of size 1920x1009

Time gen imagen: 2487 ms
Time gen imagen: 2577 ms
Time gen imagen: 2586 ms
Time gen imagen: 2587 ms
Time gen imagen: 2601 ms
Time gen imagen: 2617 ms
Time gen imagen: 2605 ms
Time gen imagen: 2573 ms
Time gen imagen: 4806 ms
Time gen imagen: 2483 ms
Time gen imagen: 3205 ms
Time gen imagen: 2834 ms
Time gen imagen: 3021 ms
Time gen imagen: 2660 ms
Time gen imagen: 6266 ms
Time gen imagen: 2660 ms
Time gen imagen: 3032 ms
```
Paladin Test Five370_DeathPvE_20240709203715.png
Paladin Test Five370_DeathPvE_20240709203715

module:

image
image
image

### **Expected**

https://github.com/mehah/otclient/assets/114332266/f56a3ed2-fb40-48f8-89e8-1255bbbea57d

only game windows / framebuffer

image

## Fixes

#707

## Type of change

  - [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested

  - [x] Test in 13.10 tfs main repo
  - [x] Only test Death and level up
  
**Test Configuration**:

  - Server Version: 13.10
  - Client: Main repo
  - Operating System: windows 10

## Checklist
  - [x] I have performed a self-review of my own code
  - [x] My changes generate no new warnings
  - [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: modules/game_screenshot/game_screenshot.lua [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    {id = ScreenshotType.PLAYER_ATTACKING, label = "Player Attacking", enableDefault = false},
+    {id = ScreenshotType.HIGHEST_DAMAGE_DEALT, label = "Highest Damage Dealt", enableDefault = false},
+    {id = ScreenshotType.HIGHEST_HEALING_DONE, label = "Highest Healing Done", enableDefault = false},
+    {id = ScreenshotType.LOW_HEALTH, label = "Low Health", enableDefault = false},
+    {id = ScreenshotType.GIFT_OF_LIFE_TRIGGERED, label = "Gift of Life Triggered", enableDefault = true}
+}
+-- LuaFormatter on
+
+-- @ widget
+local evento = nil
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: change all `evento` to `screenshotScheduleEvent`

File: modules/game_screenshot/game_screenshot.lua [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    destroyOptionsModule()
+
+    if evento then
+        removeEvent(evento)
+        evento = nil
+    end
+end
+
+function screenshotController:onGameStart()
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    for _, temp in ipairs(AutoScreenshotEvents) do
+        local label = g_ui.createWidget("ScreenshotType", optionPanel.allCheckBox)
+        local settingKey = temp.label:gsub("%s+", "")
+        local setings = g_settings.getBoolean(settingKey)
+        label.text:setText(temp.label)
+        label.enabled:setChecked(setings)
+        label.enabled:setId(temp.id)
+        temp.currentBoolean = setings
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    end
+
+    if not g_resources.directoryExists(autoScreenshotDir) then
+        g_resources.makeDir(autoScreenshotDirName)
+    end
+
+    -- is g_game or LocalPlayer ?
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this comment

File: modules/game_screenshot/game_screenshot.lua [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    end
+    if evento then
+        removeEvent(evento)
+        evento = nil
+    end
+    for _, evento in ipairs(AutoScreenshotEvents) do
+        local labelSinEspacios = evento.label:gsub("%s+", "")
+        g_settings.set(labelSinEspacios, evento.currentBoolean)
+    end
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        return
+    end
+    local id = tonumber(widget:getId())
+    for _, temp in ipairs(AutoScreenshotEvents) do
+        if temp.id == id then
+            temp.currentBoolean = checked
+            break
+        end
+    end
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    for _, i in pairs(optionPanel.allCheckBox:getChildren()) do
+        for _, j in pairs(i:getChildren()) do
+            if j:getStyle().__class == 'UICheckBox' then
+                local id = tonumber(j:getId())
+                if id then
+                    for _, evento in ipairs(AutoScreenshotEvents) do
+                        if evento.id == id then
+        --      hadouken
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this comment

File: modules/game_screenshot/game_screenshot.lua [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        --      hadouken                            
+                            j:setChecked(evento.enableDefault)
+                            break
+                        end
+                    end
+                end
+            end
+        end
+    end
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+end
+-- LuaFormatter on
+
+function destroyOptionsModule()
+    modules.client_options.removeTab('Screenshot')
+    if optionPanel and not optionPanel:isDestroyed() then
+        optionPanel:destroy()
+        optionPanel = nil
+    end
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+end
+
+function onScreenShot(type)
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        if evento.id == type and evento.currentBoolean then
+
+            local screenshotName = name .. level .. "_" .. evento.label:gsub("%s+", "") .. "_" ..
+                                       os.date("%Y%m%d%H%M%S") .. ".png"
+            takeScreenshot("/" .. autoScreenshotDirName .. "/" .. screenshotName)
+
+            return
+        end
+    end
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: remove this space

File: modules/game_screenshot/game_screenshot.lua [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+end
+function takeScreenshot(name)
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: add space

**Review by luanluciano93 — CHANGES_REQUESTED** [review]

File: modules/game_screenshot/game_screenshot.lua [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    destroyOptionsModule()
+
+    if screenshotScheduleEvent then
+        removeEvent(screenshotScheduleEvent)
+        screenshotScheduleEvent = nil
+    end
+end
+
+function screenshotController:onGameStart()
+    if g_game.getClientVersion() < 1310 then
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: `11.80 `
https://tibia.fandom.com/wiki/Updates/11.80.7048

File: modules/game_screenshot/game_screenshot.lua [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        g_resources.makeDir(autoScreenshotDirName)
+    end
+
+    screenshotController:registerEvents(LocalPlayer, {
+        onTakeScreenshot = onScreenShot
+    })
+end
+
+function screenshotController:onGameEnd()
+    if g_game.getClientVersion() >= 1310 then
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]: `11.80 `
https://tibia.fandom.com/wiki/Updates/11.80.7048

File: modules/game_screenshot/game_screenshot.lua [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+
+    for _, temp in ipairs(AutoScreenshotEvents) do
+        local label = g_ui.createWidget("ScreenshotType", optionPanel.allCheckBox)
+        local settingKey = temp.label:gsub("%s+", "")
+        local setings = g_settings.getBoolean(settingKey)
+        label.text:setText(temp.label)
+        label.enabled:setChecked(setings)
+        label.enabled:setId(temp.id)
+        temp.currentBoolean = setings
+    end
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]:
```suggestion
    for _, screenshotEvent in ipairs(AutoScreenshotEvents) do
        local label = g_ui.createWidget("ScreenshotType", optionPanel.allCheckBox)
        local settingKey = screenshotEvent.label:gsub("%s+", "")
        local settings = g_settings.getBoolean(settingKey)
        label.text:setText(screenshotEvent.label)
        label.enabled:setChecked(settings)
        label.enabled:setId(screenshotEvent.id)
        screenshotEvent.currentBoolean = settings
    end
```

File: modules/game_screenshot/game_screenshot.lua [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+        destroyOptionsModule()
+    end
+    if screenshotScheduleEvent then
+        removeEvent(screenshotScheduleEvent)
+        screenshotScheduleEvent = nil
+    end
+    for _, evento in ipairs(AutoScreenshotEvents) do
+        local labelSinEspacios = evento.label:gsub("%s+", "")
+        g_settings.set(labelSinEspacios, evento.currentBoolean)
+    end
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]:
```suggestion
    for _, screenshotEvent in ipairs(AutoScreenshotEvents) do
        local labelScreenshotEvent = screenshotEvent.label:gsub("%s+", "")
        g_settings.set(labelScreenshotEvent, screenshotEvent.currentBoolean)
    end
```

File: modules/game_screenshot/game_screenshot.lua [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+    if not widget then
+        return
+    end
+    local id = tonumber(widget:getId())
+    for _, temp in ipairs(AutoScreenshotEvents) do
+        if temp.id == id then
+            temp.currentBoolean = checked
+            break
+        end
+    end
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]:
```suggestion
    for _, screenshotEvent in ipairs(AutoScreenshotEvents) do
        if screenshotEvent.id == id then
            screenshotEvent.currentBoolean = checked
            break
        end
    end
```

File: modules/game_screenshot/game_screenshot.lua (hunk 7 above)
Comment (kokekanon) [reply modules/game_screenshot/game_screenshot.lua]: image

**Comment by kokekanon** [comment]
from my side  is ready
but
## missing content (I don't know how to do that)
image

File: modules/game_screenshot/game_screenshot.lua [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+end
+
+-- LuaFormatter off
+function resetValues()
+    for _, evento in ipairs(AutoScreenshotEvents) do
+        local labelSinEspacios = evento.label:gsub("%s+", "")
+        g_settings.set(labelSinEspacios, evento.currentBoolean)
+
+    end
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]:
```suggestion
    for _, screenshotEvent in ipairs(AutoScreenshotEvents) do
        local labelScreenshotEvent = screenshotEvent.label:gsub("%s+", "")
        g_settings.set(labelScreenshotEvent, screenshotEvent.currentBoolean)
    end
```

File: modules/game_screenshot/game_screenshot.lua [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+                    for _, evento in ipairs(AutoScreenshotEvents) do
+                        if evento.id == id then                
+                            j:setChecked(evento.enableDefault)
+                            break
+                        end
+                    end
+                end
+            end
+        end
+    end
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]:
```suggestion
    for _, selectedCheckBox in pairs(optionPanel.allCheckBox:getChildren()) do
        for _, selectedCheckBoxChildren in pairs(selectedCheckBox:getChildren()) do
            if selectedCheckBoxChildren:getStyle().__class == 'UICheckBox' then
                local id = tonumber(selectedCheckBoxChildren:getId())
                if id then
                    for _, screenshotEvent in ipairs(AutoScreenshotEvents) do
                        if screenshotEvent.id == id then                
                            selectedCheckBoxChildren:setChecked(screenshotEvent.enableDefault)
                            break
                        end
                    end
                end
            end
        end
    end
```

File: modules/game_screenshot/game_screenshot.lua [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+
+    screenshotScheduleEvent = scheduleEvent(function()
+        g_app.doScreenshot(name)
+    end, 50)
+end
+
+function OpenFolder()
+    local dir = g_resources.getWriteDir():gsub("[/\\]+", "\\") .. autoScreenshotDirName
+    g_platform.openDir(dir)
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]:
```suggestion
    local directory = g_resources.getWriteDir():gsub("[/\\]+", "\\") .. autoScreenshotDirName
    g_platform.openDir(directory)
```

File: modules/game_screenshot/game_screenshot.lua [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+    local name = g_game.getLocalPlayer():getName() or "player"
+    local level = g_game.getLocalPlayer():getLevel() or 1
+    for _, evento in ipairs(AutoScreenshotEvents) do
+        if evento.id == type and evento.currentBoolean then
+            local screenshotName = name .. level .. "_" .. evento.label:gsub("%s+", "") .. "_" ..
+                                       os.date("%Y%m%d%H%M%S") .. ".png"
+            takeScreenshot("/" .. autoScreenshotDirName .. "/" .. screenshotName)
+            return
+        end
+    end
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.lua]:
```suggestion
    for _, screenshotEvent in ipairs(AutoScreenshotEvents) do
        if screenshotEvent.id == type and screenshotEvent.currentBoolean then
            local screenshotName = name .. level .. "_" .. screenshotEvent.label:gsub("%s+", "") .. "_" ..
                                       os.date("%Y%m%d%H%M%S") .. ".png"
            takeScreenshot("/" .. autoScreenshotDirName .. "/" .. screenshotName)
            return
        end
    end
```

**Review by luanluciano93 — CHANGES_REQUESTED** [review]

File: src/client/protocolgameparse.cpp [hunk 22]
```diff
@@ -548,7 +548,11 @@ void ProtocolGame::parseMessage(const InputMessagePtr& msg)
                 case Proto::GameServerBosstiaryCooldownTimer:
                     parseBosstiaryCooldownTimer(msg);
                     break;
+                case Proto::GameServerTakeScreenshot:
+                    parseTakeScreenshot(msg);
+                    break;
 
+                    // OTC custom server communication
```
Comment (luanluciano93) [inline src/client/protocolgameparse.cpp]: ?

File: modules/game_screenshot/game_screenshot.otui [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+
+
+      
+  Button
+    !text: tr('Reset')
+    width: 64
+    anchors.right: parent.right
+    anchors.bottom: parent.bottom
+    @onClick: modules.game_screenshot.resetValues()
+  Button
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.otui:69]: add space

File: modules/game_screenshot/game_screenshot.otui [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+    width: 164
+    anchors.left: parent.left
+    anchors.bottom: parent.bottom
+    @onClick: modules.game_screenshot.OpenFolder()
+
+ScreenshotType < UIWidget
+  background-color: alpha
+  text-offset: 35 1
+  focusable: true
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.otui]: remove space

File: modules/game_screenshot/game_screenshot.otui [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+    anchors.bottom: Opciones3.bottom
+
+    layout:
+      type: grid
+      cell-size: 240 20
+      cell-spacing: 1
+      flow: true
+      fit-children: true
+
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.otui]: remove 2 spaces

File: modules/game_screenshot/game_screenshot.otui [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+    anchors.right: parent.right
+    anchors.top: prev.bottom
+    margin-top: 7
+    height: 22
+  
+    OptionCheckBoxMarked
+      id: wtfIsThis2
+      !text: tr('Keep Blacklog of the Screenshots of the Last 5 Seconds')
+      !tooltip: tr('If you check this option, the client takes a screenshot every\nsecond and caches the last 5 of them. Whenever a screenshot is\ntaken either automatically for a selected event or manually using a\nhotkey, also the screenshots of the previous 5 seconds will be\nsaved.\nIf you experience frame stuttering, you should consider disabling\nthis option..')
+
```
Comment (luanluciano93) [inline modules/game_screenshot/game_screenshot.otui:24]: remove 2 spaces

File: src/client/protocolgameparse.cpp [hunk 27]
```diff
@@ -4202,6 +4206,12 @@ void ProtocolGame::parseBosstiaryEntryChanged(const InputMessagePtr& msg) {
     msg->getU32(); // bossId
 }
 
+void ProtocolGame::parseTakeScreenshot(const InputMessagePtr& msg) {
+
```
Comment (luanluciano93) [inline src/client/protocolgameparse.cpp]: remove space

File: src/client/protocolgameparse.cpp [hunk 28]
```diff
@@ -4202,6 +4206,12 @@ void ProtocolGame::parseBosstiaryEntryChanged(const InputMessagePtr& msg) {
...
+    const uint8_t screenshotType = msg->getU8();      //Type id screenshot
```
Comment (luanluciano93) [inline src/client/protocolgameparse.cpp]:
```suggestion
    const uint8_t screenshotType = msg->getU8();
```

**Review by luanluciano93 — CHANGES_REQUESTED** [review]

**Review by majestyotbr — APPROVED** [review]

**Pull request merged**